vip-jetpack/connection-pilot/class-jetpack-connection-pilot.php

```diff
@@ -0,0 +1,128 @@
+
+			if ( ! wp_next_scheduled( self::CRON_ACTION ) ) {
+				wp_schedule_event( strtotime( sprintf( '+%d minutes', mt_rand( 1, 60 ) ) ), self::CRON_SCHEDULE, self::CRON_ACTION );
+			}
+		}
+	}
+
+	/**
+	 * The main cron job callback.
+	 * Checks the JP connection and alerts/auto-resolves when there are problems.
+	 *
+	 * Needs to be static due to how it is added to cron control.
+	 */
+	public static function run_connection_pilot() {
+		if ( ! self::should_run_connection_pilot() ) {
+			return;
+		}
+
+		require_once __DIR__ . '/class-jetpack-connection-controls.php';
+		require_once __DIR__ . '/class-jetpack-connection-status-check.php';
+
+		$status_check = new WPCOM_VIP_Jetpack_Connection_Status_Check();
+		$status_check->launch();
+
+		// Send out notifications.
+		if ( is_array( $status_check->pilot_notifications ) ) {
+			foreach ( $status_check->pilot_notifications as $notification ) {
+				self::send_alert( $notification['message'], $notification['error'], $notification['healthcheck'] );
+			}
+		}
+	}
+
+	/**
+	 * Send an alert to IRC and Slack.
+	 *
+	 * Example message:
+	 * Jetpack is disconnected, but was previously connected under the same domain.
+	 * Site: example.go-vip.co (ID 123). The last known connection was on August 25, 12:11:14 UTC to Cache ID 65432 (example.go-vip.co).
+	 * Jetpack connection error: [jp-cxn-pilot-not-active] Jetpack is not currently active.
+	 *
+	 * @param string   $message optional.
+	 * @param WP_Error $wp_error optional.
+	 * @param array    $last_healthcheck optional.
+	 *
+	 * @return mixed True if the message was sent to IRC, false if it failed. If sandboxed, will just return the message string.
+	 */
+	private static function send_alert( $message = '', $wp_error = null, $last_healthcheck = null ) {
+		$message .= sprintf( ' Site: %s (ID %d).', get_site_url(), defined( 'VIP_GO_APP_ID' ) ? VIP_GO_APP_ID : 0 );
+
+		if ( isset( $last_healthcheck['site_url'], $last_healthcheck['cache_site_id'], $last_healthcheck['last_healthcheck'] ) ) {
+			$message .= sprintf(
+				' The last known connection was on %s UTC to Cache ID %d (%s).',
+				date( 'F j, H:i', $last_healthcheck['last_healthcheck'] ), $last_healthcheck['cache_site_id'], $last_healthcheck['site_url']
+			);
+		}
+
+		if ( is_wp_error( $wp_error ) ) {
+			$message .= sprintf( ' Jetpack connection error: [%s] %s', $wp_error->get_error_code(), $wp_error->get_error_message() );
+		}
+
+		if ( ( defined( 'WPCOM_SANDBOXED' ) && WPCOM_SANDBOXED ) || ( ! defined( 'ALERT_SERVICE_ADDRESS' ) ) ) {
+			return $message; // Just return the message, as posting to IRC won't work.
+		}
+
+		return wpcom_vip_irc( '#vip-jp-cxn-monitoring', $message );
+	}
+
+	/**
+	 * Checks if the connection pilot should run.
+	 *
+	 * @return bool True if the connection pilot should run.
+	 */
+	private static function should_run_connection_pilot() {
+		if ( defined( 'WPCOM_VIP_RUN_CONNECTION_PILOT' ) && WPCOM_VIP_RUN_CONNECTION_PILOT ) {
```

Comment:
this should have an explicit true check as well

Also - if someone manually runs this (like I did) without the define being set, they may not realize it's necessary. I wonder what's the best way to detect/warn about that?

And do we need the ability to run this without scheduling an event?

Reply:
Ideally, there will be CLI commands with the shared functionality in the future. ATM, jetpack-start is sufficient when needing to do manual tests and fixes though. So I'm not sure if this is the right place to optimize/change.